modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java

```diff
@@ -1298,6 +1319,93 @@ private CompletableFuture<Boolean> onPrimaryExpired(PrimaryReplicaEventParameter
             return falseCompletedFuture();
         }
 
+        // TODO: move to Replica https://issues.apache.org/jira/browse/IGNITE-23750
+        private void registerFailoverCallback(TablePartitionId replicationGroupId) {
+            CompletableFuture<Replica> replicaFuture = replicaManager.replica(replicationGroupId);
+
+            // TODO: https://issues.apache.org/jira/browse/IGNITE-23753
+            // For now we assume that it's a possible situation to get null future on node's stopp process, but it should be fixed with
+            // a correct and consistent replica's state.
+            if (replicaFuture == null) {
+                return;
+            }
+
+            replicaFuture.thenApply(electedPrimaryReplica -> {
+                onLeaderElectedFailoverCallback = (leaderNode, term) -> changePeersAndLearnersAsyncIfPendingExists(
+                        electedPrimaryReplica,
+                        replicationGroupId,
+                        term
+                );
+
+                return electedPrimaryReplica.raftClient().subscribeLeader(onLeaderElectedFailoverCallback);
+            }).exceptionally(e -> {
+                LOG.error("Rebalance failover subscription on elected primary replica failed [groupId=" + replicationGroupId + "].", e);
+
+                return null;
+            });
+        }
+
+        // TODO: move to Replica https://issues.apache.org/jira/browse/IGNITE-23750
+        private void unregisterFailoverCallback(PrimaryReplicaEventParameters parameters) {
+            CompletableFuture<Replica> replicaFuture = replicaManager.replica(parameters.groupId());
+
+            // TODO: https://issues.apache.org/jira/browse/IGNITE-23753
+            // For now we assume that it's a possible situation to get null future on node's stopp process, but it should be fixed with
+            // a correct and consistent replica's state.
+            if (replicaFuture == null) {
+                return;
+            }
+
+            replicaFuture.thenAccept(expiredPrimaryReplica -> expiredPrimaryReplica.raftClient()
+                    .unsubscribeLeader(onLeaderElectedFailoverCallback)
+            );
+        }
+
+        private void changePeersAndLearnersAsyncIfPendingExists(
+                Replica primaryReplica,
+                TablePartitionId replicationGroupId,
+                long term
+        ) {
+            getPendingAssignmentsSupplier.apply(replicationGroupId).exceptionally(e -> {
+                LOG.error(
+                        format(
```

Comment:
There's no need in using format, there's a `error(String msg, @Nullable Throwable th, Object... params)` method.

Reply:
Done!